### ISSUE TITLE
Add MINIMAL example of how to deploy on Google Cloud

### ISSUE THREAD

**Issue by bgoodman44** [dask/dask-kubernetes#177]
I personally am having a bit of trouble deploying my dask workload on GKE.  I doubt I'm the only one, as the current "quickstart" guide makes a lot of assumptions and has a lot of extra stuff (helm, jupyterhub) that confuses the core ideas for those not familiar with gce deployments or kubernetes. 

I suggest adding a minimal deployment tutorial that includes:

-  Setting up a gke cluster (already covered by zero-to-jupyterhub)

- What about local ssds?  Does each pod require a local ssd if the underlying datasets exceed worker memory.  Or, does the dask-worker-space reside on the "master" node?  Do the pods have access the master nodes disk?

- Creating/linking a docker image for the worker
- Can I build the image on the "master" node and the workers will be able to find it?
-  Writing a yaml file for the gke cluster and docker image (not covered in any detail, from what I can find).

Here is the toy script I'm working with to try to get running on GKE:

```
from dask.distributed import Client,get_worker
from dask_kubernetes import KubeCluster

def write_file(idx):
    fName = "file_{:04d}.txt".format(idx);
    content = fName + " " + str(get_worker().id) + "\n";
    with open(fName,'w+') as f:
        f.write(content);
    return (fName,content);

if __name__=="__main__":
    cluster = KubeCluster();
    client = Client(cluster);
    futs = []
    for ii in range(0,10):
        futs.append( client.submit(write_file,ii) );

    print(futs,'\n\n')
    out = client.gather(futs);
    print(out)

```

Dockerfile:

```
FROM python:3.6
RUN mkdir /app
WORKDIR /app
ADD . /app/
RUN pip install -r requirements.txt
```

Any help would be appreciated!

**Comment by bgoodman44** [comment]
I should note:  The above script will error out because the "worker pod specification not provided".  If you remove the KubeCluster part...it will work locally.

**Comment by jacobtomlinson** [comment]
Thanks for raising this @bgoodman44! Having clearer documentation would also be great. As a new user to the project, you are going to be in a better position than most to help us improve it.

I'm not super concerned about the helm and JupyterHub references as they are optional. Perhaps it could be made even clearer that they are optional though?

Here are some direct responses to your questions. Let me know your thoughts and feel free to ask more. 

> Setting up a gke cluster (already covered by zero-to-jupyterhub)

I'm hesitant to duplicate existing documentation on how to set up a GKE cluster here. There are many ways of creating a Kubernetes cluster, using GKE isn't going to suit everyone, and this potentially opens us up to writing and maintaining many different guides on this. Perhaps as an alternative, we could link to existing documentation on this. For example, [the official Kubernetes docs](https://kubernetes.io/docs/setup/production-environment/turnkey/gce/) already has some excellent guides.

> What about local ssds? Does each pod require a local ssd if the underlying datasets exceed worker memory. Or, does the dask-worker-space reside on the "master" node? Do the pods have access the master nodes disk?

If your workers run out of memory they will spill to local disk. This is local to the pod on the host machine. An advanced configuration could include setting up SSD volumes for this space to accelerate things a little, but it should just work out of the box anyway.

> Creating/linking a docker image for the worker

What could we add to the "Docker Images" section to help answer this question? The quickstart example shows using the official Dask image from docker hub which is public and will be pulled to the workers automatically by Kubernetes.

You could optionally build your own image and either push it to docker hub and update the image name in the worker spec. Or you could push it to a private registry like the one on GCE, but this would likely take some more configuration in GKE to allow it to pull the image.

> Can I build the image on the "master" node and the workers will be able to find it?

If you are suggesting running `docker build ...` on one of the Kubernetes nodes directly then no it will not be found by the other nodes automatically. You need to push it into a registry that is accessible. Again for the quickstart, it may be easier to leave the default settings and use the official Dask image.

> Writing a yaml file for the gke cluster and docker image (not covered in any detail, from what I can find).

Again this is Google Cloud specific and ideally the quickstart should be generic. What yaml file are you asking out specifically here?

> The above script will error out because the "worker pod specification not provided". If you remove the KubeCluster part...it will work locally.

Yes this is because you need to update your dask config with your worker spec. The quickstart shows an example of placing this in a file and then creating your cluster by passing the config explicitly. You could also put this in your `~/.config/dask/kubernetes.yaml` config file under the `kubernetes.worker-template-path` key, this is already noted in the "Configuration" section. How could we improve this to make it clearer?

**Comment by Techn0logic** [comment]
I agree, the use cases will be different for everyone. We want to deploy adaptive dask clusters on kubernetes in a cloud provider agnostic fashion, for multiple users, complex environments, very large workloads etc. Currently mostly developing on AWS custom clusters. 

We have some custom packages that compute on dask with some tricky and complex conda environments.

Our workflow is creating virtualenv which contains all dependencies with `conda-pack` on build image > copying environment onto application image > pushing to docker registry > making sure worker-speck.yaml points to that image > running KubeCluster > running client.compute. It's important to make sure that `dask-kubernetes` and `distributed` versions on your notebook env and worker env are the same.

So we rely on `docker push` to make sure that the environment on the worker image is in sync with what scientists/developers expect to avoid serialization problems. 

Our other use case for example is to use `dask-kubernetes` on kubeflow https://github.com/kubeflow/kubeflow

**I really thing that the documentation should be specific to `dask-kubernetes` only. As the package can and will be used in many different scenarios and deployments.**

I think the only bit that is missing (or I didn't find it) in the `quickstart` is setting up kubernetes roles and permissions to the notebook pod to allow listing/starting and killing worker pods. It would be great to have a summary of expected roles and rolebindings with some code snippets and kubernetes commands. This would make it clear how to ensure that the scheduler can interact with worker pods, and worker pods can communicate back to the scheduler. 

And also, the documentation seems to be geared towards starting the scheduler from inside the k8s cluster on the notebook. Or at least it's recommended - however, the notebooks interface is super tricky for production grade development and complex packages. Users should be able to easily request adaptive compute from their dev machines, ide's, qa envs, production etc.

**Comment by jacobtomlinson** [comment]
> It would be great to have a summary of expected clusterroles and rolebindings with some code snippets and kubernetes commands.

Completely agree with this. Are you interested in contributing this?

> the documentation seems to be geared towards starting the scheduler from inside the k8s cluster on the notebook.

Currently the scheduler must be within the cluster, or at least the cluster network fabric much be routable from wherever you are and vice-versa. Either in a Python script or a notebook, but maybe the docs are a little heavy on the notebook. This will be fixed as of #162 where the `KubeCluster` can be created anywhere (as you can get to a loadbalancer, nodeport, etc) and the scheduler will be created as a pod within the cluster.

**Comment by bgoodman44** [comment]
I certainly sympathize with wanting to remain generic, but in my opinion, the current quick start is NOT accomplishing this, since it assumes that your environment is already up and running.

I'm happy to put together a complete/verbose/minimal "for dummies" (like me) quickstart, but I have to get my workload working on GKE first.  In my opinion, something like this should exist for the dominant cloud providers.  Sort of like how the zero-to-jupyterhub tutorial branches and goes into detail for each of the major providers.

**Comment by jacobtomlinson** [comment]
I certainly take your point but I would argue that complete/verbose != minimal/quickstart. 

I would definitely be happy to review and merge a longer guide with more detail, but I don't think the quickstart is the right place for it.

**Comment by bgoodman44** [comment]
Fair enough.  I'll reach out when I have something to share

Would you prefer I close this or keep it open?

**Comment by Techn0logic** [comment]
> > It would be great to have a summary of expected clusterroles and rolebindings with some code snippets and kubernetes commands.
> 
> Completely agree with this. Are you interested in contributing this?

Yes - happy to do that. 

> the documentation seems to be geared towards starting the scheduler from inside the k8s cluster on the notebook.
> 
Really looking forward to #162 


With regards to guides for GKE, AWS EKS or Kubernetes on AWS EC2 etc. I haven't used managed kubernetes from cloud providers (such as GKE and AWS EKS), but I think once the kubernetes cluster is ready, then deploying `dask-kubernetes` won't be any different to deploying over minikube, or EC2 kubernetes deployments.

Looks like it's already mentioned in here already https://docs.dask.org/en/latest/setup/kubernetes-helm.html with links to resources on how to set up k8s cluster, on cloud providers or minikube

**Comment by jacobtomlinson** [comment]
@bgoodman44 happy to leave open.

**Comment by jacobtomlinson** [comment]
@bgoodman44 have you had any chance to take a look at this again?

**Comment by bgoodman44** [comment]
I've got an outline, but I'm currently stuck on RBAC. Haven't figured out how to allow the pod that the KubeCluster is launched in to create new worker pods

**Comment by jacobtomlinson** [comment]
You need to create a service account for that pod with the correct RBAC perms. We have [this config](https://github.com/pangeo-data/helm-chart/blob/master/pangeo/templates/dask-kubernetes-rbac.yaml) over in the [Pangeo project](http://pangeo.io/) which adds dask-kubernetes support to [Zero to JupyterHub](https://zero-to-jupyterhub.readthedocs.io/en/latest/).

**Comment by bgoodman44** [comment]
Thank you Jacob!  It works and it is GLORIOUS!  I still have a few things to hash out, but than I'll share my "how-to".

**Comment by manujchandra** [comment]
@bgoodman44 Hi, did you manage to setup a working adaptive cluster ok GKE? If yes, could you please share the steps? Thanks.
